view::linear_distribute

gnzlbg opened the pull request:
Adds `view::linear_distribute(from, to, n)` which linearly distributes `n`
elements in the closed range `[from, to]`. It models `ForwardRange`.

Its semantics are:

- Precondition: `from <= to and n > 0`.
- Requirements: `std::is_arithmetic<T>`.
- Returns:
  - If `n == 1`, it returns `to`.
  - If `from == to`, it returns `n`-times to.
  - Otherwise, distirbutes `n` equally-spaced points in the interval
  `[from, to]` (the end-points are always included).

Possible future extensions:

- Support user-defined integer types: see issues #605 ande #591.
- Model `BidirectionalRange`.
- Model `RandomAccessRange` for `std::is_integral<T>`.

Closes #102.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+/// \file
```
This header needs to be added to `range/v3/view.hpp`.

gnzlbg replied:
fixed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+            {
+                return from_;
+            }
+            constexpr void next()
```
`RANGES_CXX14_CONSTEXPR`

gnzlbg replied:
fixed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+            {
+                return from_ > to_;
+            }
+            constexpr void prev()
```
`RANGES_CXX14_CONSTEXPR`

gnzlbg replied:
fixed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+            {
+                from_ -= spacing_;
+            }
+            constexpr void advance(std::ptrdiff_t n)
```
`RANGES_CXX14_CONSTEXPR`

gnzlbg replied:
fixed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+            }
+            constexpr void advance(std::ptrdiff_t n)
+            {
+                RANGES_EXPECT(to_ - from_ >= n * spacing_);
```
Maybe `RANGES_EXPECT(std::numeric_limits<T>::max() / spacing_ >= n);` before this?

gnzlbg replied:
indeed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+            }
+        public:
+            constexpr linear_distribute_view() = default;
+            constexpr linear_distribute_view(T from, T to, std::size_t n)
```
Either `RANGES_CXX14_CONSTEXPR`, or encode the body as comma expressions in the member-initializer-list, a la
```c++
constexpr linear_distribute_view(T from, T to, std::size_t n)
  : from_((RANGES_EXPECT(n > 1), detail::move(from))),
    to_((
        RANGES_EXPECT(to > from_),
        RANGES_EXPECT((to - from_) / static_cast<T>(n - 1) > T{0}),
        detail::move(to)))
{}
```

Love those C++11 `constexpr` hacks.

gnzlbg replied:
Fixed: I went with `RANGES_CXX14_CONSTEXPR`, nice workaround though.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,123 @@
+                template<typename T,
+                    CONCEPT_REQUIRES_(std::is_arithmetic<T>())>
+                constexpr auto operator()(T from, T to, std::size_t n) const
+                    RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT(linear_distribute_view<T>{from, to, n})
```
Typical style is to not indent `RANGES_DECLTYPE_AUTO_XXX`, and to put the parens where braces would go in a "real" function body:
```c++
constexpr auto operator/*...*/
RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
(
    linear_distribute_view<T>{from, to, n}
)
```

gnzlbg replied:
fixed

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,127 @@
+            }
+            RANGES_CXX14_CONSTEXPR void advance(std::ptrdiff_t n)
+            {
+                RANGES_EXPECT(std::numeric_limits<T>::max() / spacing_ >= n);
```
`#include<limits>`

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,127 @@
+            CONCEPT_REQUIRES(std::is_floating_point<T>())
+            constexpr std::ptrdiff_t distance_to(linear_distribute_view const &that) const
+            {
+                return static_cast<std::ptrdiff_t>(std::floor((that.from_ - from_) / spacing_));
```
`#include<cmath>`

ericniebler replied:
I think this is a `ForwardRange`. Otherwise, the corrections for errors at each step gives you different answers. I suggest removing `prev`, `advance`, and `distance_to`, at least for floating point types.

gnzlbg replied:
@ericniebler I've stick to a `ForwardRange` for all types now. Relaxing this to a `BidirectionalRange` turned out to be tricky to get right for floating point types.

CaseyCarter commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,35 @@
+
+  using ranges::view::linear_distribute;
+
+  auto float_eq = [](double a, double b) { return (int)std::floor(std::abs(a - b)) == 0; };
```
`#include<cmath>`

gnzlbg commented:
The failures here also look spurious, @ericniebler could you attempt to retry the failed jobs please?

ericniebler commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,129 @@
+            }
+            RANGES_CXX14_CONSTEXPR void next()
+            {
+                from_ += spacing_;
```
From Sean (emphasis mine):
> There are many ways to deal with the error at each step, two common ways are error accumulation and stochastic rounding. A simple method of error accumulation is to subtract off each segment **and re-divide the remaining part into the remaining number of segments**.

You're not doing it right. ;-) I think this is closer to Sean's intent:

```c++
from_ += (to_ - from_) / n_--;
```

... where `n_` is the (integral) number of segments. You're done when `n_` is zero.

gnzlbg replied:
pinging @sean-parent . The implementation provided supports `RandomAccess`, has no divide in the tight loop, and produces the same results as Matlab's `linspace` function.

I think it is a mistake to enable `RandomAccess` for floating point values (i'll fix that) since `from_ = n * spacing_` and ` while (n > 0) from_ += spacing` won't produce the same result due to floating point errors, but with a constant spacing it can still be a `BidirectionalRange`.

gnzlbg replied:
So i've refactored the view into one for floating point numbers and one for integers, and for the floats one I've implemented it like @sean-parent suggested, so its not random access anymore. We can tweak it later to something different if we feel like it in a smaller PR.

ericniebler commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,129 @@
+              RANGES_EXPECT(n > 1);
+              RANGES_EXPECT(to_ > from_);
+              spacing_ = (to_ - from_)/ static_cast<T>(n - 1);
+              RANGES_EXPECT(spacing_ > T{0});
```
Spacing.

gnzlbg replied:
fixed

gnzlbg commented:
I still need to look into why the floating point range is not bidirectional.

ericniebler commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,60 @@
+                             2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9,
+                             3.0},
+                            float_eq));
+    }
```
You need to test the `to==from && n==1` case, for both integral and floating point types. And I'm curious what happens in the `to==from && n>1` case.

gnzlbg replied:
done

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                }
+                constexpr bool equal(default_sentinel) const
+                {
+                  return spacing_ == T{0}? from_ != to_ : from_ == to_ + spacing_;
```
Couldn't this simply `return from_ > to_;`?

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                }
+                RANGES_CXX14_CONSTEXPR void prev()
+                {
+                    from_ -= spacing_;
```
Doesn't this need `if (spacing_ == T{0}) { --to_; RANGES_EXPECT(to_ == from_); }` to mirror `next`?

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                }
+                RANGES_CXX14_CONSTEXPR void advance(std::ptrdiff_t n)
+                {
+                    
```
`advance(1)` seems to not be equivalent to `next()` when `spacing_ == 0`.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                    from_ += n * spacing_;
+                }
+                constexpr std::ptrdiff_t distance_to(linear_distribute_integral_view const &that) const
+                {
```
`RANGES_EXPECT(spacing_ != T{0});`?

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,183 @@
+                    else
+                    {
+                        RANGES_EXPECT(n > 1);
+                        spacing_ = (to_ - from_)/ static_cast<T>(n - 1);   
```
> You can replace the `if`/`else` with:

If `to_ == from_` and `n == 1`, that divides by zero.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                linear_distribute_integral_view(T from, T to, std::ptrdiff_t n)
+                  : from_(from), to_(to) 
+                {
+                    RANGES_EXPECT(n > 0);
```
This is redundant with the checks in the `if` / `else`, and `to >= from_` is partially redundant. Why not replace the whole body with:
```c++
if (to_ == from_)
{
    RANGES_EXPECT(n == 1);
    spacing_ = 0;
}
else
{
    RANGES_EXPECT(n > 1);
    RANGES_EXPECT(from < to_);
    spacing = (to_ - from_) / static_cast<T>(n - 1);
}
```

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+            struct linear_distribute_floating_point_view
+              : view_facade<linear_distribute_floating_point_view<T>, finite>
+            {
+                static_assert(std::is_floating_point<T>(), "");
```
Ditto `CONCEPT_ASSERT`

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,183 @@
+                constexpr bool equal(default_sentinel) const
+                {
+                    return n_ == 0;
+                }
```
> You need to add an `equal` that compares...

...in order to model `Sentinel<I, I>()` and `EqualityComparable<I>()`. Is there any motivation for doing so in addition to modeling `InputIterator<I>()` and `Sentinel<default_sentinel, I>()`?

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+                }
+                RANGES_CXX14_CONSTEXPR void prev()
+                {
+                    from_ -= (to_ - from_) / T(++n_);
```
I suspect there are corner cases where this does not perfectly invert the behavior of `next`. No correction yet suggested, but we need to be careful here.

gnzlbg replied:
Indeed, this cannot work. @ericniebler 

I am making the range `ForwardRange` for now. Making it `Bidirectional` is a bit trickier, might do that in a follow-up PR.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,183 @@
+                }
+            public:
+                linear_distribute_floating_point_view() = default;
+                linear_distribute_floating_point_view(T from, T to, std::ptrdiff_t n)
```
and `noexcept`.

gnzlbg replied:
fixed.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+        /// Distributes `n >= 2` values linearly in the closed interval [`from`, `to`].
+        template <typename T>
+        using linear_distribute_view
+        = meta::if_c<std::is_floating_point<T>::value,
```
Either this line needs an indent, or the next two need outdents.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,183 @@
+                {
+                    CONCEPT_ASSERT_MSG(std::is_arithmetic<T>(),
+                        "The object passed to view::linear_distribute must model the Arithmetic "
+                        "concept; that is, it must be one of the native integer of floating-point types");
```
Do we need to perform that relaxation now, before accepting this PR?

gnzlbg replied:
I'm with Casey on this, we should have a solid concept for user-defined "integers" before trying to bolt it on here. Having a couple of views that require `std::is_integral` (or `std::is_arithmetic`), will help getting that concept right.

CaseyCarter commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,80 @@
+        CHECK(ranges::size(irng) == std::size_t{2});
+        CHECK(ranges::at(irng, 0) == 0);
+        CHECK(ranges::at(irng, 1) == 1);
+        static_assert(ranges::RandomAccessRange<decltype(irng)>(), "");
```
`CONCEPT_ASSERT` all of these, please.

gnzlbg replied:
done

gnzlbg commented:
So this was all wrong. In Matlab, there are two different types of these kinds of ranges:

1. `linspace`: what is now `view::linear_distribute`, with the following semantics: 
  - [`x1`, `x2`]: Point interval, specified as a pair of numeric scalars. `x1` and `x2` define the interval over which linspace generates points. `x1` and `x2` can be real or complex, and `x2` can be either larger or smaller than `x1`. If `x2` is smaller than `x1`, then the vector contains descending values.
 - `n`: Number of points, specified as a real numeric scalar. If `n` is `1`, linspace returns `x2`. If `n` is zero or negative, linspace returns an empty 1-by-0 matrix. If `n` is not an integer, linspace rounds down and returns `floor(n)` points.

2. The colon operator (`:`): `x = j:i:k`creates a regularly-spaced vector `x` using `i` as the increment between elements. The vector elements are roughly equal to `[j,j+i,j+2*i,...,j+m*i]` where `m = fix((k-j)/i)`. However, if `i` is not an integer, then floating point arithmetic plays a role in determining whether colon includes the endpoint `k` in the vector, since `k` might not be exactly equal to `j+m*i`. If you specify nonscalar arrays, then MATLAB interprets `j:i:k` as `j(1):i(1):k(1)`.

I've now chosen for `view::linear_distribute` to stick to the semantics of `linspace`. The only change is that it does not support "descending" ranges, since in range-v3 one can just use `view::reverse` to opt into that. 

The colon operator `:` should probably belong into something like `view::linear_step`, where instead of the number of points, one specifies the step to use.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,115 @@
+                RANGES_CXX14_CONSTEXPR void next()
+                {
+                    RANGES_EXPECT(n_ > 0);
+                    if (n_ == 1)
```
`if (--n_ == 0)`, and remove the decrements in the branches?

gnzlbg replied:
I've moved the decrement outside the conditional.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,115 @@
+            struct linear_distribute_view
+              : view_facade<linear_distribute_view<T>, finite>
+            {
+                static_assert(std::is_arithmetic<T>(), "");
```
Please `CONCEPT_ASSERT(foo)` all of these `static_assert(foo, "")` declarations.

gnzlbg replied:
fixed here and in the .cpp file

gnzlbg commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,60 @@
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014
```
fixed

gnzlbg commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,60 @@
+#include "../simple_test.hpp"
+#include "../test_utils.hpp"
+
+int main() {
```
fixed

gnzlbg commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,191 @@
+            struct linear_distribute_integral_view
+              : view_facade<linear_distribute_integral_view<T>, finite>
+            {
+                static_assert(std::is_integral<T>(), "");
```
fixed

gnzlbg commented:
So I've simplified this as much as possible, it now only models `ForwardRange`, anything else is left as an extension to follow up PRs. I've updated the PR info and the commit message and rebased.

sean-parent commented:
It has been a _long_ time since I've done any significant work in this space, but providing some info that may help. linear_distribute is a simple case of a linear recurrence algorithm. These are very important in many domains, including graphics and physics. In graphics, a linear_distribution is the answer to "how do you draw a line". The problem with doing this with a discrete machine is how to do error propagation. The simple way is to simply recalculate at each step:
```cpp
template <class Out>
void iterative(int f, int l, unsigned n, Out out) {
    while (n != 0) {
        f += (l - f) / n;
        --n;
        *out++ = f;
    }
    assert((n == 0) || (f == l));
}
```
(just using int here as a placeholder). This gives a very simple solution to error distribution. To do this quickly, you can use forward differencing. Bresenham's algorithm for drawing a line is a forward differencing algorithm to do linear distribution. Another common approach is to recursively subdivide. Recursive subdivision has a couple of advantages over forward differencing, it typically requires far fewer bits of precision at each step, and it allows you to drill to any location within the solution in log(n) steps. The downside is higher cost at each step and log(n) storage. However, if you want a "random access" linear distribution, this is how to get it. The recursive form of the above code is:
```cpp
template <class Out>
void recursive(int f, int l, int n, Out out) {
    if (n == 0) return;
    if (n & 1) {
        f += (l - f) / n;
        --n;
        *out++ = f;
    }
    auto m = f + (l - f) / 2;

    recursive(f, m, n / 2, out);
    recursive(m, l, n / 2, out);
}
```
Of course, one can use hybrid solutions between the two. There is a _lot_ of work that has been done in this space for graphics. The same basic approach is used for drawing nearly all graphics primitives (lookup how to draw cubic and quadratic Bézier curves). In Physics you see the same approaches to plot trajectories at a given time intervals (as one example), and to find collision points.

The same basic algorithm applies to distributing floating point values, you simply have a smaller epsilon in that case. Note that iterators behave enough like integers that you can (and it is very useful to) have a distribution of iterators.

CaseyCarter commented on include/range/v3/utility/iterator.hpp:
```diff
@@ -189,7 +189,7 @@ namespace ranges
                 RANGES_EXPECT((Same<I, S>() || 0 <= n));
                 D d = bound - it;
-                RANGES_EXPECT(0 <= n ? 0 <= d : 0 >= d);
+                //RANGES_EXPECT(0 <= n ? 0 <= d : 0 >= d);
```
Commenting out a precondition check is literally *never* the right thing.

gnzlbg replied:
oops, this shouldn't haven't been commented out. I had a lot of debugging logic in there when i was trying to reverse engineer what `distance_to(view const& other)` actually was supposed to do when doing the RandomAccess version.

CaseyCarter commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,116 @@
+                    return n_ == 0;
+                }
+                constexpr bool equal(linear_distribute_view const& other) const noexcept
+                {
```
`RANGES_EXPECT(from_ == other.from_ && to_ == other.to_)`?

gnzlbg replied:
I guess now is the time to ask my questions about `equal(view const& other)` and `distance_to(view const& other)` when are these actually called and for what? 

If they are only called for sub-views of the same range, then the assertion is correct, but if they can be called for completely independent ranges, then that should be part of the result.

CaseyCarter replied:
`equal(view const&)` is called by `operator==(basic_iterator<view> const&, basic_iterator<view> const&)`. You can support (slowly) comparing iterators from different ranges by returning `std::tie(from_, to_, n_) == std::tie(that.from_, that.to_, that.n_)`, or you can declare that only iterators from the same range are in the domain of equality and return `(RANGES_EXPECT(std::tie(from_, to_) == std::tie(that.from_, that.to_)), n_ == that.n_)` or `(RANGES_EXPECT(std::tie(to_, n_) == std::tie(that.to_, that.n_)), from_ == that.from_)`. The flavor that compares only a single integer will probably be the fastest in release builds. In the name of "don't pay for what you don't use," range-v3 always chooses speed for comparison with ranges over enlarging the domain of `==`. (It would be awesome if STL2 were to have distinct hooks for "fast comparison" using operators, and "total ordering" using some other customization point so that we could usefully provide both operations on iterators.)

`distance_to(view const& that)` is called by `operator-(basic_iterator<view> const&, basic_iterator<view> const&)`. Similar reasoning applies here, although I think you would have a much harder time making this operation work across distinct ranges than `equal`.

gnzlbg replied:
That makes it clear. I will do this for integers but will omit the `RANGES_EXPECT` for floating point values (since comparing for equality there does not make sense without some sort of tolerance).

CaseyCarter commented on test/view/linear_distribute.cpp:
```diff
@@ -0,0 +1,82 @@
+    {
+        auto irng = linear_distribute(0, 21, 22);
+        auto frng = linear_distribute(0., 21., 22);
+        check_equal(irng, frng | ranges::view::transform([](double v) { return int(v); }));
```
`[](double v) { return int(v); }` could be `ranges::convert_to<int>{}` from `functional.hpp`.

sean-parent commented:
One more comment - there is more than one way to distribute error in a linear distribution. For example, if you are a drawing a linear gradient (a linear distribution of colors) than often it is better to use a stochastic model where the accumulated error is a probability of the position. This avoids visible rounding artifacts.

gnzlbg commented:
@sean-parent thanks for the pointers. The current implementation is using the simple solution you mention, I'll try to extend it to recursive subdivision, since I think the `log(N)` stack space cost is worth it for `RandomAccess`. 

I think that at least for floating point values it is important to document the error distribution being used. 

Using a stochastic model would probably belong into a different `view`. I don't really see an easy way of allowing the user to customize the error distribution without basically having to re-implement the whole `view`.

sean-parent commented:
My algorithms also only handle the case where the range is larger than n correctly (technically, the distance(f,l) measured in epsilons must not be < n). Otherwise iterative and recursive differ and you get a pile-up with iterative at the beginning). The normal way to deal with that in graphics is to flip coordinates (there is inherent symmetry here, a linear distribution is a line (f,0),(l, n) when distance(f,l) < n then you can swap x and y and use the same algorithm to get proper distribution. Algorithmically it is useful to have a distribution of 0 elements (returning nothing). Don't know if that fits in the view model.

ericniebler commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,124 @@
+                constexpr bool equal(linear_distribute_view const& other) const noexcept
+                {
+                    return n_ == other.n_;
+                }
```
Why keep these two `equal` overloads separate instead of combining them with an `EXPECT` like:

```
RANGES_EXPECT(std:: is_floating_point <T>()  || (from_ == other.from_ && to_ == other.to_));
 ```

OK, so the floating point case is no longer C++11 `constexpr`, but I don't think that matters.

gnzlbg replied:
Because floating point comparison with `==` (even if it is never invoked) did raise warnings (unless one uses `if constexpr`). This was the "nicest" solution I found :/

CaseyCarter replied:
> the floating point case is no longer C++11 constexpr

*Neither* case would be C++11 constexpr, although that's easily fixed by combining `RANGES_EXPECT(/**/)` and `n_ == other.n_` into a comma expression.

CaseyCarter replied:
> Because floating point comparison with == (even if it is never invoked) did raise warnings

Far from expert on anything to do with fp, but I think the warnings are spurious here. If we cannot consistently and exactly repeat the values of `from_` and `to_` on repeated passes over the same range - and I think this implementation correctly does so - then we are violating the iterator requirements and we need to fix it.

ericniebler replied:
Agreed. Copies of floating point values should be equality comparable, and floating point operations are deterministic and equality preserving to the best of my knowledge. (I hate to think of the implications if they weren't.)

I think one `equal` that does `RANGES_EXPECT(from_ == other.from_ && to_ == other.to_)` with a warning suppression would be best.

ericniebler commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,124 @@
+                    }
+                }
+            public:
+                RANGES_CXX14_CONSTEXPR linear_distribute_view() noexcept = default;
```
s/`noexcept = default`/`= default`/

ericniebler commented on include/range/v3/view/linear_distribute.hpp:
```diff
@@ -0,0 +1,118 @@
+                RANGES_CXX14_CONSTEXPR
+                bool equal(linear_distribute_view const& other) const noexcept
+                {
+                    RANGES_EXPECT(from_ == other.from_ && to_ == other.to_);
```
This needs a warning suppression.

```
/Users/travis/build/ericniebler/range-v3/include/range/v3/view/linear_distribute.hpp:52:41: error: comparing floating point with == or != is unsafe [-Werror,-Wfloat-equal]
                    RANGES_EXPECT(from_ == other.from_ && to_ == other.to_);
                                  ~~~~~ ^  ~~~~~~~~~~~
```

CaseyCarter replied:
> This needs a warning suppression.

https://github.com/gnzlbg/range-v3/pull/6

gnzlbg commented:
travis is all green

ericniebler commented:
Thanks for all your work on this, @gnzlbg !

skgbanga commented:
Ah! That's very informative. Thanks!
